timemachine/optimize/protocol.py

```diff
@@ -208,7 +223,142 @@ def max_work_stddev_distance(prev_lam, next_lam, max_step=0.25):
     return max_work_stddev_distance
 
 
-def greedily_optimize_protocol(distance_fxn: DistanceFxn, target_distance=0.5, max_iterations=1000) -> Array:
+# (2) pair overlap
+def reweighted_pair_overlap(u_n_A, u_n_B, u_n_ref):
+    """given arrays [u(x) for x in xs] (for u in {u_A, u_B, u_ref}),
+    estimate overlap(A, B) by reweighting from ref->A and ref->B
+
+    Notes
+    -----
+    * see sec. 3.4 of https://pubmed.ncbi.nlm.nih.gov/25808134/ for expression being approximated
+    * TODO: describe approximation approach
+    """
+    # reduced potentials -> unnormalized log probs
+    log_q_A = -u_n_A
+    log_q_B = -u_n_B
+    log_q_ref = -u_n_ref
+    log_N = jnp.log(len(log_q_ref))
+
+    # pick a normalization constant for ref
+    log_p_ref_n = log_q_ref - logsumexp(log_q_ref - log_N)
+
+    # normalize A and B wrt ref
+    log_p_A = log_q_A - logsumexp(log_q_A - log_p_ref_n - log_N)
+    log_p_B = log_q_B - logsumexp(log_q_B - log_p_ref_n - log_N)
+
+    log_prod_AB_n = log_p_A + log_p_B
+    # assert (log_prod_AB_n < jnp.inf).all()
+
+    log_p_mix_n = logsumexp(jnp.array([log_p_A, log_p_B]), axis=0)
+    # assert (log_p_mix_n < jnp.inf).all()
+
+    log_denom = log_p_mix_n + log_p_ref_n
+    mask = log_denom > -jnp.inf  # mask out div by 0
+    log_f_n = jnp.where(mask, log_prod_AB_n - log_denom, 0.0)
+    # assert not jnp.isnan(log_f_n).any()
+
+    log_mean_f = logsumexp(log_f_n - jnp.log(sum(mask)))
+    overlap = jnp.exp(log_mean_f)
+    return 2 * overlap  # so that it goes to 1.0 when u_n_A == u_n_B
+
+
+def make_one_minus_similarity_fxn(sim_fxn):
+    def one_minus_f(a, b):
+        return 1 - sim_fxn(a, b)
+
+    return one_minus_f
+
+
+def make_overlap_fxn(u_lam, src_u_n):
+    """Compose a reduced potential function u_lam with the reweighting approximation from reweighted_pair_overlap
+
+    Parameters
+    ----------
+    u_lam : function : scalar -> array of length N
+        u_lam(lam) = [u(xs[n], lam) for n in range(N)]
+
+    src_u_n : array
+        src_u_n = [u_src(xs[n]) for n in range(N)]
+
+    Returns
+    -------
+    overlap_fxn: (lam_a, lam_b) -> overlap_estimate
+    """
+
+    def overlap_fxn(lam_a, lam_b):
+        estimate = reweighted_pair_overlap(u_lam(lam_a), u_lam(lam_b), src_u_n)
+        clamped = jnp.clip(estimate, 0.0, 1.0)  # TODO: revert this line if estimates > 1 are avoided
+        return clamped
+
+    return overlap_fxn
```

Comment:
Nice! I like how this is defined so that the user can provide $u(\lambda$ (e.g. if the linear approximation is insufficient?), but using a common code path.

Reply:
Exactly -- if the user is concerned about the quality of the default recommended approximation (linearly interpolate axis 0 of  pre-computed `u_kn` given `lambda_k`), the user can supply a different implementation.

I shied away from supplying these implementations in the current PR, but the case of single topology intermediates in vacuum, for example, might look like:

```python
def make_reference_u_lam_vacuum(st: SingleTopology, trajs: List[Trajectory]):
	xs = np.vstack([traj.frames for traj in trajs])
	kBT = BOLTZ * temperature

	def ref_u_lam(lam):
		U_lam_fxn = st.setup_intermediate_state(lam).get_U_fn()
		return vmap(U_lam_fxn)(xs) / kBT
	
	return ref_u_lam
```

Perhaps there is an appropriate place to define such functions for all common cases (`SingleTopology` in {vacuum, solvent, complex}, absolute in solvent, ...), but initially seems best to leave that as a detail that can be optionally supplied by a user.